sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/AnsiTypeCoercion.scala

```diff
@@ -0,0 +1,305 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.catalyst.analysis
+
+import scala.annotation.tailrec
+
+import org.apache.spark.sql.catalyst.analysis.TypeCoercion.numericPrecedence
+import org.apache.spark.sql.catalyst.expressions._
+import org.apache.spark.sql.catalyst.expressions.aggregate._
+import org.apache.spark.sql.catalyst.plans.logical.LogicalPlan
+import org.apache.spark.sql.catalyst.rules.Rule
+import org.apache.spark.sql.types._
+
+/**
+ * In Spark ANSI mode, the type coercion rules are based on the type precedence lists of the input
+ * data types.
+ * As per the section "Type precedence list determination" of "ISO/IEC 9075-2:2011
+ * Information technology - Database languages - SQL - Part 2: Foundation (SQL/Foundation)",
+ * the type precedence lists of primitive data types are as following:
+ *   * Byte: Byte, Short, Int, Long, Decimal, Float, Double
+ *   * Short: Short, Int, Long, Decimal, Float, Double
+ *   * Int: Int, Long, Decimal, Float, Double
+ *   * Long: Long, Decimal, Float, Double
+ *   * Decimal: Any wider Numeric type
+ *   * Float: Float, Double
+ *   * Double: Double
+ *   * String: String
+ *   * Date: Date, Timestamp
+ *   * Timestamp: Timestamp
+ *   * Binary: Binary
+ *   * Boolean: Boolean
+ *   * Interval: Interval
+ * As for complex data types, Spark will determine the precedent list recursively based on their
+ * sub-types.
+ *
+ * With the definition of type precedent list, the general type coercion rules are as following:
+ *   * Data type S is allowed to be implicitly cast as type T iff T is in the precedence list of S
+ *   * Comparison is allowed iff the data type precedence list of both sides has at least one common
+ *     element. When evaluating the comparison, Spark casts both sides as the tightest common data
+ *     type of their precedent lists.
+ *   * There should be at least one common data type among all the children's precedence lists for
+ *     the following operators. The data type of the operator is the tightest common precedent
+ *     data type.
+ *       * In
+ *       * Except(odd)
+ *       * Intersect
+ *       * Greatest
+ *       * Least
+ *       * Union
+ *       * If
+ *       * CaseWhen
+ *       * CreateArray
+ *       * Array Concat
+ *       * Sequence
+ *       * MapConcat
+ *       * CreateMap
+ *   * For complex types (struct, array, map), Spark recursively looks into the element type and
+ *     applies the rules above. If the element nullability is converted from true to false, add
+ *     runtime null check to the elements.
+ *  Note: this new type coercion system will allow implicit converting String type literals as other
+ *  primitive types, in case of breaking too many existing Spark SQL queries.
```

Comment:
The standard says something about this behaviour, too?

Reply:
No, this is not from the standard. It is learned from PostgreSQL. I planed to put this feature in another follow-up PR. But there were multiple test failures from the "SQLQueryTestSuite", so I did it in this PR as well.
What do you think of it?

Reply:
I just updated the comment to claim that it is not from the standard.

Reply:
Also, this is for passing TPCDSQueryANSISuite.

E.g. without promoting string literals, q83 will fail with
```
org.apache.spark.sql.AnalysisException: cannot resolve '(spark_catalog.default.date_dim.`d_date` IN ('2000-06-30', '2000-09-27', '2000-11-17'))' due to data type mismatch: Arguments must be same type but were: date != string; line 12 pos 17;
```